tests/simulator/py/fluid_init.py

```diff
@@ -0,0 +1,244 @@
+        ratios = {key: noise[0] * np.sqrt(nppc[0]) for key, noise in noises.items()}
+        self.assertTrue(
+            all([diff < 0.09 for diff in _get_diff(ratios, nppc, noises).values()])
+        )
+        for key, noise in noises.items():
+            _save_std_fig(
+                nppc, noise, ratios[key], fn_name + key, test_threashold=0.007
+            )
+        _save_dbg_fig(datasets, model_dict, fn_name)
+
+    def checkPopulationDensityEqualsUserFunction(self, fluid_diags, interp):
+        fn_name = "density"
+        sim = ph.globals.sim
+        model_dict = ph.globals.sim.model.model_dict
+
+        pop_densities = _pop_name_to_patch_level0(
+            fluid_diags, FluidInitValidation.pop_diag_file(fn_name)
+        )
+
+        def user_density(pop_name, component_name, x):
+            return model_dict[pop_name][component_name](x)
+
+        datasets, std_devs = _get_std_devs_for(
+            sim, pop_densities, model_dict, user_density
+        )
+        dev_per_ppc = _sort_std_devs_per_ppc(pop_densities, std_devs, model_dict)
+        noises = _get_noise_from(dev_per_ppc)
+        nppc = [x[1] for x in dev_per_ppc]
+        ratios = {key: noise[0] * np.sqrt(nppc[0]) for key, noise in noises.items()}
+        self.assertTrue(_get_diff(ratios, nppc, noises)[fn_name] < 0.09)
+        _save_std_fig(nppc, noises[fn_name], ratios[fn_name], fn_name)
+        _save_dbg_fig(datasets, model_dict, fn_name)
+
+
+def _pop_name_to_patch_level0(diags, Type):
+    pop_patch_level = {}
+    for diag in selectDiags(diags, Type):
+        patches = list(diag.levels[0].patches.values())
+        assert len(patches)
+        # pull "alpha" from hd5f path eg /t#/pl0/p0#0/ions/pop/ions_alpha/${fn_name}
+        pop_name = patches[0].patch_data.h5item.name.split("/")[-2][5:]
+        pop_patch_level[pop_name] = diag.levels[0]
+
+    return pop_patch_level
+
+
+def _get_std_devs_for(sim, pop_patch_levels, model_dict, lmbda):
+    datasets, std_devs = {}, {}
+    for pop_name, patch_level in pop_patch_levels.items():
+        patches = list(patch_level.patches.values())
+        std_devs[pop_name] = {}
+        cell_width = patch_level.cell_width("x")
+        datasets[pop_name] = aggregate_level0_patch_domain(
+            patch_level, shared_patch_border=True
+        )
+        for data_name in patch_level.data_names():
+            x = np.arange(sim.cells[0] + 1) * cell_width + sim.origin[0]
+            std_devs[pop_name][data_name] = np.std(
+                lmbda(pop_name, data_name, x) - datasets[pop_name][data_name]
+            )
+
+    return datasets, std_devs
+
+
+def _sort_std_devs_per_ppc(pop_patch_levels, std_devs, model_dict):
+    return sorted(
+        [
+            (std_devs[pop_name], model_dict[pop_name]["nbrParticlesPerCell"])
+            for pop_name in list(pop_patch_levels.keys())
+        ],
+        key=lambda tple: tple[1],
+    )
+
+
+def _get_noise_from(devs_per_ppc):
```

Comment:
I don't get what you mean by "noise", and I have no idea what keys from ` devs_per_ppc[0][0].keys()` are, which means I don't really get what this function is doing

Reply:
```
def _get_noise_from(devs_per_ppc):
    """
      devs_per_ppc = list[(data_name: std_dev, n_ppc)] # sorted per ppc
    """
    noises = {}
    for data_name_dev_dict, _ in devs_per_ppc:
        for data_name, std_dev in data_name_dev_dict.items():
            if data_name not in noises:
                noises[data_name] = []
            noises[data_name].append(std_dev)
    return noises # = data_name:list[std_dev] # list sorted by ppc ascending
```